### PR TITLE
Do merge ntfs & calendar

### PR DESCRIPTION
**Pull request by datanel**
Handle ntfs or calendar file
Merge if necessary

### CODE REVIEW

File: tartare/core/calendar_handler.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def _join_calendar_lines(calendar_lines, lines):
     for calendar_line in calendar_lines:
         for line in lines:
             if calendar_line['network_id'] == line['network_id']:
-                if '' == calendar_line['line_code'] or line['line_code'] == calendar_line['line_code']:
+                if (not 'line_code' in calendar_line) or '' == calendar_line['line_code'] or line['line_code'] == calendar_line['line_code']:
```
Comment (antoine-de) [inline tartare/core/calendar_handler.py]: I think you can rewrite this:
`if not calendar_line.get('line_code') or line['line_code'] == calendar_line['line_code']:`

File: tartare/default_settings.py [hunk 2]
```diff
@@ -15,6 +15,8 @@
 
 CELERY_DEFAULT_EXCHANGE = 'celery_tartare'
 
+CELERYD_CONCURRENCY = 1
```
Comment (antoine-de) [inline tartare/default_settings.py]: I don't think this should be done, but if you want, we can fix it later (but in this case I think it should be nice to put a comment)

File: tartare/tasks.py [hunk 3]
```diff
@@ -95,12 +110,35 @@ def remove_old_ntfs_files(directory):
...
 
+
 def create_dir(directory):
     """create directory if needed"""
     if not os.path.exists(directory):
         os.makedirs(directory)
 
 
+def _do_merge_calendar(calendar_file, ntfs_file, output_file):
+    calendars_zip = ZipFile(calendar_file, 'r')
```
Comment (antoine-de) [inline tartare/tasks.py]: I think it's better to use a context manager here (to ensure file closure):

`with ZipFile(calendar_file, 'r') as calendar_zip:`

thus the `done()` is done automatically

File: tartare/tasks.py [hunk 4]
```diff
@@ -95,12 +110,35 @@ def remove_old_ntfs_files(directory):
...
+    new_ntfs_zip = calendar_handler.merge_calendars_ntfs(grid_calendar_data, ntfs_zip)
+    calendar_handler.save_zip_as_file(new_ntfs_zip, output_file)
+    calendars_zip.close()
+    ntfs_zip.close()
+
+
+def _get_current_nfts_file(current_data_dir):
+    files = glob.glob(os.path.join(current_data_dir, "*"))
+
+    for f in files:
```
Comment (antoine-de) [inline tartare/tasks.py]: I think you can rewrite this:
`next((f for f in files if os.path.isfile(f) and f.endswith('.zip') and is_calendar_data(f)), None)`

File: tartare/tasks.py [hunk 5]
```diff
@@ -111,10 +149,36 @@ def handle_data(input_dir, output_dir, current_data_dir):
...
-            #NTFS file is moved to the CURRENT_DATA_DIR, old NTFS file is deleted
+            # NTFS file is moved to the CURRENT_DATA_DIR, old NTFS file is deleted
             remove_old_ntfs_files(current_data_dir)
             copyfile(input_file, os.path.join(current_data_dir, filename))
-        shutil.move(input_file, output_file)
+            if os.path.isfile(calendar_file):
+                # Merge
+                _do_merge_calendar(calendar_file, input_file, output_file)
+                os.remove(input_file)
+            else:
```
Comment (antoine-de) [inline tartare/tasks.py]: we move it if it's not a file ? (we want to move dir ?)

File: tests/integration/rest_api_test.py [hunk 6]
```diff
@@ -50,7 +50,7 @@ def test_post_grid_calendar_returns_success_status(app):
     files = {'file': (open(path, 'rb'), 'export_calendars.zip')}
     raw = app.post('/grid_calendar', data=files)
     r = to_json(raw)
-    backup_dir = os.path.join(tartare.app.config.get("GRID_CALENDAR_DIR"))
+    backup_dir = os.path.join(tartare.app.config.get("INPUT_DIR"))
```
Comment (antoine-de) [inline tests/integration/rest_api_test.py]: you can thus rename the variable no ?

File: tartare/core/calendar_handler.py (hunk 1 above)
Comment (datanel) [inline tartare/core/calendar_handler.py]: you are not sure?

File: tests/integration/update_data_test.py [hunk 7]
```diff
@@ -32,5 +34,84 @@ def test_handle_interesting_data(tmpdir):
...
+    calendar_file = os.path.join(os.path.dirname(os.path.dirname(__file__)),
+                                 'fixtures/gridcalendar/export_calendars.zip')
+    copy(calendar_file, str(calendar_dir))
+
+    handle_data(str(input), str(output), str(current_dir))
+
+    files_in_output_dir = os.listdir(str(output))
+
+    assert files_in_output_dir[0].endswith('ntfs.zip')
+    assert os.path.isfile(str(current_dir.join('ntfs.zip')))
```
Comment (antoine-de) [inline tests/integration/update_data_test.py]: shouldn't we check that the calendars have been updated ?

File: tests/integration/update_data_test.py [hunk 8]
```diff
@@ -32,5 +34,84 @@ def test_handle_interesting_data(tmpdir):
...
+    input = tmpdir.mkdir('input')
+    calendar_file = os.path.join(os.path.dirname(os.path.dirname(__file__)),
+                                 'fixtures/gridcalendar/export_calendars.zip')
+    copy(calendar_file, str(input))
+    output = tmpdir.mkdir('output')
+    current_dir = tmpdir.mkdir('current_dir')
+
+    handle_data(str(input), str(output), str(current_dir))
+
+    assert os.path.isfile(str(current_dir.join('grid_calendar').join('export_calendars.zip')))
```
Comment (antoine-de) [inline tests/integration/update_data_test.py]: maybe we can check that the file is not in outputdir too ?

File: tests/integration/update_data_test.py [hunk 9]
```diff
@@ -32,5 +34,84 @@ def test_handle_interesting_data(tmpdir):
...
+    ntfs_zip = ZipFile(os.path.join(str(current_dir), 'ntfs.zip'), 'a', ZIP_DEFLATED, False)
+    for filename in glob(ntfs_path):
+        ntfs_zip.write(filename, os.path.basename(filename))
+    ntfs_zip.close()
+
+    handle_data(str(input), str(output), str(current_dir))
+
+    files_in_output_dir = os.listdir(str(output))
+
+    assert files_in_output_dir[0].endswith('ntfs.zip')
```
Comment (antoine-de) [inline tests/integration/update_data_test.py]: same here, wouldn't it be nice to check that the calendars have been updated

File: tests/integration/update_data_test.py (hunk 8 above)
Comment (datanel) [inline tests/integration/update_data_test.py]: done

**Comment by datanel** [comment]
Jenkins can you test this please?

**Pull request merged**